### PR TITLE
tmpdir: prevent symlink attacks and TOCTOU races (CVE-2025-71176)

### LINKED ISSUE #13669: Vulnerable tmpdir handling
Starting around https://github.com/pytest-dev/pytest/blob/main/src/_pytest/tmpdir.py#L156, there are some heuristics to make sure that the tmpdir is safe:

```python
            temproot = Path(from_env or tempfile.gettempdir()).resolve()
            user = get_user() or "unknown"
            # use a sub-directory in the temproot to speed-up                                                                                                  
            # make_numbered_dir() call                                                                                                                         
            rootdir = temproot.joinpath(f"pytest-of-{user}")
            try:
                rootdir.mkdir(mode=0o700, exist_ok=True)
            except OSError:
                # getuser() likely returned illegal characters for the platform, use unknown back off mechanism                                                
                rootdir = temproot.joinpath("pytest-of-unknown")
                rootdir.mkdir(mode=0o700, exist_ok=True)
            # Because we use exist_ok=True with a predictable name, make sure                                                                                  
            # we are the owners, to prevent any funny business (on unix, where                                                                                 
            # temproot is usually shared).                                                                                                                     
            # Also, to keep things private, fixup any world-readable temp                                                                                      
            # rootdir's permissions. Historically 0o755 was used, so we can't                                                                                  
            # just error out on this, at least for a while.                                                                                                    
            uid = get_user_id()
            if uid is not None:
                rootdir_stat = rootdir.stat()
                if rootdir_stat.st_uid != uid:
                    raise OSError(
                        f"The temporary directory {rootdir} is not owned by the current user. "
                        "Fix this and try again."
                    )
```

This is vulnerable to symlink attacks, and probably TOCTOU attacks (see e.g. https://owasp.org/www-community/vulnerabilities/Insecure_Temporary_File). The linux kernel has protections for some of these in the `fs.protected_hardlinks` and `fs.protected_symlinks` sysctls, but they violate POSIX and are not enabled by default (though most distros enable them). In any case, they are linux-only.

For a proof of concept, create a symlink as some other user from `/tmp/pytest-of-youruser` to a directory you own (this will require `fs.protected_symlinks` to be disabled). Run pytest as yourself, and it will start writing junk to the directory that the attacker chose. It will also change the permissions on the directory he chose.

Obviously whoever wrote this was aware of these issues, but maybe not of how hard they can be to fix completely. This can be made a little bit better by telling `stat()` not to follow symlinks, but I would instead recommend getting a random base directory in a secure way (using the python tempdir functions) rather than using a predictable name.
There are just too many ways it can go wrong.

### PR DESCRIPTION
**Pull request by laurac8r**
## Summary

Replace the **predictable** `pytest-of-<user>` rootdir with a **randomly-named** rootdir created via `tempfile.mkdtemp` to prevent the entire class of predictable-name attacks (symlink races, DoS via pre-created files/dirs). As defense-in-depth, open the rootdir with `os.open` using `O_NOFOLLOW` and `O_DIRECTORY` flags and perform ownership/permission checks via fd-based `fstat`/`fchmod` to eliminate TOCTOU races.

Fixes CVE-2025-71176.

closes #13669

## Changes

- **Randomly-named rootdir**: Use `tempfile.mkdtemp(prefix=f"pytest-of-{user}-", dir=temproot)` instead of a fixed `pytest-of-{user}` directory, making pre-creation attacks infeasible.
- **`_safe_open_dir` context manager**: Opens the rootdir with `O_NOFOLLOW | O_DIRECTORY` (where available), yields the fd, and guarantees cleanup. Symlinks are rejected at the `os.open` level.
- **fd-based ownership & permission checks**: `fstat` and `fchmod` operate on the open file descriptor, eliminating the TOCTOU window between `stat`/`chmod` and the actual directory.
- **`_cleanup_old_rootdirs`**: Garbage-collects stale randomly-named rootdirs from previous sessions, respecting the retention count. The current session's rootdir is never removed.
- **Comprehensive tests**: Symlink rejection, wrong-owner rejection, missing `O_NOFOLLOW`/`O_DIRECTORY` fallback, predictable-path DoS immunity, `_cleanup_old_rootdirs` behavior, `_safe_open_dir` fd lifecycle, config validation, and session-finish edge cases.

---

- [ ] Include documentation when adding new features.
- [x] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

If this change fixes an issue, please:

- [x] Add text like ``closes #XYZW`` to the PR description and/or commits (where ``XYZW`` is the issue number). See the [github docs](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more information.

> [!IMPORTANT]
> **Unsupervised agentic contributions are not accepted**. See our [AI/LLM-Assisted Contributions Policy](https://github.com/pytest-dev/pytest/blob/main/CONTRIBUTING.rst#aillm-assisted-contributions-policy).

- [x] If AI agents were used, they are credited in `Co-authored-by` commit trailers.

Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [x] Create a new changelog file in the `changelog` directory, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/main/changelog/README.rst) for details.

  Write sentences in the **past or present tense**, examples:

  * *Improved verbose diff output with sequences.*
  * *Terminal summary statistics now use multiple colors.*

  Also make sure to end the sentence with a `.`.

- [x] Add yourself to `AUTHORS` in alphabetical order.

### CODE REVIEW

File: changelog/13669.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Fixed a symlink attack vulnerability (CVE-2025-71176) in the :fixture:`tmp_path` fixture's base directory handling.
+
+The ``pytest-of-<user>`` directory under the system temp root is now opened with ``O_NOFOLLOW`` and verified using file-descriptor-based ``fstat``/``fchmod``, preventing symlink attacks and TOCTOU races.
```
Comment (webknjaz) [inline changelog/13669.bugfix.rst]: Mind linking this abbrebiation to the definition?

File: testing/test_tmpdir.py [hunk 2]
```diff
@@ -619,3 +619,149 @@ def test_tmp_path_factory_fixes_up_world_readable_permissions(
...
+    is rejected to prevent symlink attacks."""
+    monkeypatch.setenv("PYTEST_DEBUG_TEMPROOT", str(tmp_path))
+
+    # Pre-create a target directory that the symlink will point to.
+    attacker_dir = tmp_path / "attacker-controlled"
+    attacker_dir.mkdir(mode=0o700)
+
+    # Figure out what rootdir name pytest would use, then replace it
+    # with a symlink pointing to the attacker-controlled directory.
+    import getpass
```
Comment (webknjaz) [inline testing/test_tmpdir.py]: Why is the import happening inside a function?

File: testing/test_tmpdir.py [hunk 3]
```diff
@@ -619,3 +619,149 @@ def test_tmp_path_factory_fixes_up_world_readable_permissions(
...
+    result = pytester.runpytest(p)
+    result.assert_outcomes(passed=1)
+
+
+def test_pytest_sessionfinish_handles_missing_basetemp_dir(
+    tmp_path: Path,
+) -> None:
+    """Cover the branch where basetemp is set but the directory no longer
+    exists when pytest_sessionfinish runs (314->320 partial branch)."""
+    from _pytest.tmpdir import pytest_sessionfinish
```
Comment (webknjaz) [inline testing/test_tmpdir.py]: Misplaced imports are normally discouraged. Any justification?

File: changelog/13669.bugfix.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Fixed a symlink attack vulnerability (CVE-2025-71176) in the :fixture:`tmp_path` fixture's base directory handling.
```
Comment (webknjaz) [inline changelog/13669.bugfix.rst]: There's a `:cve:` role that the `sphinx-issues` extension provides. Plus, feel free to take the credit:
```suggestion
Fixed a symlink attack vulnerability (:cve:`2025-71176`) in the :fixture:`tmp_path` fixture's base directory handling -- by :user:`laurac8r`.
```

File: changelog/13669.bugfix.rst
Comment (webknjaz) [inline changelog/13669.bugfix.rst:1]: It'd be useful to symlink `14279.bugfix.rst` to `13669.bugfix.rst` so that both are linked @ https://pytest--14279.org.readthedocs.build/en/14279/changelog.html#bug-fixes (you can also use this link to preview the change note render after pushing updates to this PR).

File: src/_pytest/tmpdir.py
Comment (webknjaz) [inline src/_pytest/tmpdir.py:1]: Some of this should probably go under a `@contextlib.contextmanager` for maintainability.

File: testing/test_tmpdir.py [hunk 5]
```diff
@@ -619,3 +619,149 @@ def test_tmp_path_factory_fixes_up_world_readable_permissions(
 
     # After - fixed.
     assert (basetemp.parent.stat().st_mode & 0o077) == 0
+
+
+@pytest.mark.skipif(not hasattr(os, "getuid"), reason="checks unix permissions")
```
Comment (webknjaz) [inline testing/test_tmpdir.py]: You can assign this to a variable and then reuse it in all these tests like so:
```python
skip_if_no_getuid = pytest.mark.skipif(not hasattr(os, "getuid"), reason="checks unix permissions")
```
```suggestion
@skip_if_no_getuid
```

File: testing/test_tmpdir.py [hunk 6]
```diff
@@ -619,3 +619,149 @@ def test_tmp_path_factory_fixes_up_world_readable_permissions(
...
+    )
+    result = pytester.runpytest(p)
+    result.assert_outcomes(passed=1)
+
+
+def test_pytest_sessionfinish_handles_missing_basetemp_dir(
+    tmp_path: Path,
+) -> None:
+    """Cover the branch where basetemp is set but the directory no longer
+    exists when pytest_sessionfinish runs (314->320 partial branch)."""
```
Comment (webknjaz) [inline testing/test_tmpdir.py]: These line numbers will become wrong over time.

File: testing/test_tmpdir.py [hunk 7]
```diff
@@ -619,3 +619,149 @@ def test_tmp_path_factory_fixes_up_world_readable_permissions(
...
+    # Point _basetemp at a path that does not exist on disk.
+    factory._basetemp = tmp_path / "already-gone"
+
+    class FakeSession:
+        class config:
+            _tmp_path_factory = factory
+
+    # exitstatus=0 + policy="failed" + _given_basetemp=None enters the
+    # cleanup block; basetemp.is_dir() is False so rmtree is skipped.
+    pytest_sessionfinish(FakeSession, exitstatus=0)
```
Comment (webknjaz) [inline testing/test_tmpdir.py:819]: Feels like this should be isolated, no?

**Comment by orlitzky** [comment]
This certainly looks like an improvement, but it's still possible for any user to DoS the machine until a reboot with `for x in $(cat /etc/passwd | cut -f1 -d:); do touch /tmp/pytest-of-$x; done;`.

File: changelog/13669.bugfix.rst
Comment (laurac8r) [reply changelog/13669.bugfix.rst:1]: oh neato, thx Sviatoslav! I'll take a look.

File: changelog/13669.bugfix.rst (hunk 1 above)
Comment (laurac8r) [reply changelog/13669.bugfix.rst]: on it!

File: src/_pytest/tmpdir.py
Comment (laurac8r) [reply src/_pytest/tmpdir.py:1]: taking a look!

File: testing/test_tmpdir.py (hunk 5 above)
Comment (laurac8r) [reply testing/test_tmpdir.py]: oh snap Sviatoslav, that's a really neat trick! love it, TY. I'm aligned.

File: changelog/13669.bugfix.rst
Comment (laurac8r) [reply changelog/13669.bugfix.rst:1]: oh, done, looks like lol! New feature of Github, huh

File: testing/test_tmpdir.py (hunk 2 above)
Comment (laurac8r) [reply testing/test_tmpdir.py]: great callout Sviatoslav! On it

File: testing/test_tmpdir.py (hunk 3 above)
Comment (laurac8r) [reply testing/test_tmpdir.py]: nope, bad habits. Nice catch! I'll search my PR for any lingering issue as such.

File: testing/test_tmpdir.py (hunk 6 above)
Comment (laurac8r) [reply testing/test_tmpdir.py]: on it!

File: testing/test_tmpdir.py (hunk 7 above)
Comment (laurac8r) [reply testing/test_tmpdir.py:819]: taking a look!

Comment (laurac8r) [reply testing/test_tmpdir.py:819]: fixed!

**Comment by laurac8r** [comment]
> This certainly looks like an improvement, but it's still possible for any user to DoS the machine until a reboot with `for x in $(cat /etc/passwd | cut -f1 -d:); do touch /tmp/pytest-of-$x; done;`.

Great callout Michael! This attack route is now blocked by a safety check before creating the directory: If a non-directory file is blocking the path (e.g. placed by another user), we attempt to remove that file and then retry.

**Comment by orlitzky** [comment]
> Great callout Michael! This attack route is now blocked by a safety check before creating the directory: If a non-directory file is blocking the path (e.g. placed by another user), we attempt to remove that file and then retry.

This cannot work unfortunately: the file is owned by the attacker, so I cannot delete his files from `/tmp` unless I am root. I indirectly mentioned this in https://github.com/pytest-dev/pytest/issues/13669#issuecomment-3218203897 where I pointed out that it's silly to tell me to fix the situation.

I don't want to sound like an asshole always coming back with new caveats to every approach, but the truth is that this is very old class of vulnerability and all of the ways to exploit it are well-known. (I was not auditing the code, I just saw the fixed path one day and knew immediately that it would be exploitable.) The solution is to use a carefully-created randomly-named path; so if someone proposes a solution that does not involve `mkstemp()` or something like it under the hood, then it's very easy to look up what the problem was with that approach.

**Comment by laurac8r** [comment]
> > Great callout Michael! This attack route is now blocked by a safety check before creating the directory: If a non-directory file is blocking the path (e.g. placed by another user), we attempt to remove that file and then retry.
> 
> This cannot work unfortunately: the file is owned by the attacker, so I cannot delete his files from `/tmp` unless I am root. I indirectly mentioned this in [#13669 (comment)](https://github.com/pytest-dev/pytest/issues/13669#issuecomment-3218203897) where I pointed out that it's silly to tell me to fix the situation.
> 
> I don't want to sound like an asshole always coming back with new caveats to every approach, but the truth is that this is very old class of vulnerability and all of the ways to exploit it are well-known. (I was not auditing the code, I just saw the fixed path one day and knew immediately that it would be exploitable.) The solution is to use a carefully-created randomly-named path; so if someone proposes a solution that does not involve `mkstemp()` or something like it under the hood, then it's very easy to look up what the problem was with that approach.

Fixed!

Previously, TempPathFactory created a fixed, predictable directory (`/tmp/pytest-of-<user>`) as the rootdir, which enabled a class of attacks: an adversary could pre-create files or symlinks at that path to DoS pytest or exploit `TOCTOU` races (CVE-2025-71176).

I've replaced `_try_ensure_directory()` with `tempfile.mkdtemp()`, which atomically creates a randomly-suffixed directory (`pytest-of-<user>-XXXX`) that cannot be predicted or pre-empted by an attacker. `os.chmod(0o700)` is applied unconditionally after `mkdtemp`, and the existing `fd`-based `fchmod` inside `_safe_open_dir` provides defense-in-depth against `TOCTOU`.

Because `mkdtemp` now produces a fresh directory every session, add `_cleanup_old_rootdirs()` to prune stale `rootdir`s from previous sessions, retaining only the `keep` most recent (by `mtime`).

**Comment by orlitzky** [comment]
> Fixed!

The new approach looks good to me FWIW.

[mkdtemp() guarantees](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp) that the directory will be 0700 and owned by the current user, so the pre-existing chmod/stat safeguards conveniently become redundant if this approach is adopted :)

**Comment by laurac8r** [comment]
> the pre-existing chmod/stat safeguards conveniently become redundant if this approach is adopted :)

TY Michael, and let's leave it to another enthusiastic coder ;) and close the book on this CVE

**Comment by orlitzky** [comment]
In light of [another recent tmpdir exploit related to directory cleanup](https://cdn2.qualys.com/advisory/2026/03/17/snap-confine-systemd-tmpfiles.txt), I think we could bullet-proof the cleanup routine even more by ensuring that https://docs.python.org/3/library/shutil.html#shutil.rmtree.avoids_symlink_attacks is true before running `rmtree()`.

It would be nice if we could confirm that the user/owner agree on the directories to be removed, and if we could remove them by file descriptor (to avoid TOCTOU again), but I don't see an obvious way to do that.

File: src/_pytest/tmpdir.py [hunk 8]
```diff
@@ -37,6 +38,70 @@
...
+                Path(entry.path)
+                for entry in os.scandir(temproot)
+                if entry.is_dir(follow_symlinks=False)
+                and entry.name.startswith(prefix)
+                and Path(entry.path) != current
+            ),
+            key=lambda p: p.lstat().st_mtime,
+            reverse=True,
+        )
+    except OSError:
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py:117]: Why this try/except? Please add a comment for future reference.

File: src/_pytest/tmpdir.py [hunk 9]
```diff
@@ -190,6 +258,8 @@ def getbasetemp(self) -> Path:
                 lock_timeout=LOCK_TIMEOUT,
                 mode=0o700,
             )
+            # Clean up old rootdirs from previous sessions.
+            _cleanup_old_rootdirs(temproot, rootdir_prefix, keep, current=rootdir)
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: I noticed that before we used to keep 3 directories around, but on this branch we always end up with 4 directories, which is a bit confusing if configured to keep 3 directories around (the default).

File: src/_pytest/tmpdir.py [hunk 10]
```diff
@@ -37,6 +38,70 @@
 RetentionType = Literal["all", "failed", "none"]
 
 
+@contextlib.contextmanager
+def _safe_open_dir(path: Path) -> Generator[int]:
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py:39]: This function is only used in one place:

https://github.com/pytest-dev/pytest/blob/bb1371b0547db2de200910f068e7b7b841a18478/src/_pytest/tmpdir.py#L242-L254

I think we should instead have a more focused function: `_verify_ownership_and_tighthen_permission(path, user_id)`, instead of splitting the logic that we ultimately want in two separate places.

File: testing/test_tmpdir.py [hunk 11]
```diff
@@ -539,6 +550,100 @@ def test_on_rm_rf_error(self, tmp_path: Path) -> None:
...
+
+        import shutil
+
+        monkeypatch.setattr(shutil.rmtree, "avoids_symlink_attacks", True)
+
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            safe_rmtree(target)
+        symlink_warnings = [x for x in w if "avoids_symlink_attacks" in str(x.message)]
+        assert len(symlink_warnings) == 0
```
Comment (nicoddemus) [inline testing/test_tmpdir.py]: This lets us see the warning in case something goes wrong:

```suggestion
        symlink_warnings = [x for x in w if "avoids_symlink_attacks" in str(x.message)]
        assert symlink_warnings == []
```

File: testing/test_tmpdir.py [hunk 12]
```diff
@@ -539,6 +550,100 @@ def test_on_rm_rf_error(self, tmp_path: Path) -> None:
         assert not fn.is_file()
 
 
+class TestSafeRmtree:
+    """Tests for safe_rmtree and the avoids_symlink_attacks guard."""
```
Comment (nicoddemus) [inline testing/test_tmpdir.py]:
```suggestion
    """Tests for safe_rmtree and the avoids_symlink_attacks guard (#13669)."""
```

File: testing/test_tmpdir.py [hunk 13]
```diff
@@ -592,30 +697,344 @@ def test_tmp_path_factory_create_directory_with_safe_permissions(
...
     tmp_factory = TempPathFactory(None, 3, "all", lambda *args: None, _ispytest=True)
     basetemp = tmp_factory.getbasetemp()
 
-    # After - fixed.
+    # Should still create the directory with safe permissions.
+    assert basetemp.is_dir()
     assert (basetemp.parent.stat().st_mode & 0o077) == 0
+
+
+def test_tmp_path_factory_from_config_rejects_negative_count(
```
Comment (nicoddemus) [inline testing/test_tmpdir.py]: I think we might be going overboard with the number of tests here. Could you verify if it is possible to remove/merge some of them?

File: src/_pytest/tmpdir.py (hunk 9 above)
Comment (nicoddemus) [reply src/_pytest/tmpdir.py]: Another thing I noticed is now that we always generate a "root" temporary dir on every run:

`pytest-of-Bruno-xe7su5xh`, `pytest-of-Bruno-vwyhe35x`, etc, which is fine as it is the whole point of this PR.

However, each directory also has a `pytest-0` and `pytest-current` (a symlink to `pytest-0`), which no longer makes sense in this scenario.

File: changelog/13669.bugfix.rst [hunk 14]
```diff
@@ -0,0 +1,9 @@
+Fixed a symlink attack vulnerability ([CVE-2025-71176](https://github.com/pytest-dev/pytest/issues/13669)) in
```
Comment (nicoddemus) [inline changelog/13669.bugfix.rst]: This description is detailed, but we should begin it with the user facing change:

```rst
Temporary directories created by :fixture:`tmp_path` used to be created inside a `pytest-of-{user}` root directory within the system's temporary directory. Each session would create a new subdirectory with an incrementing counter -- `pytest-0`, `pytest-1`, etc. -- along with a `pytest-current` symlink pointing to the current one.

This left the system vulnerable to symlink attacks, so pytest has changed how these directories are created.

Now, each session creates a single `pytest-of-{user}-{random}` directory directly inside the system's temporary directory. This makes the directories unpredictable and safe from symlink attacks and `TOCTOU <https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use>`__ races.
```

Also note that this entry should be written in ReStructuredText, not Markdown.

**Comment by salmankadaya** [comment]
Could you please let me know when the new release containing this fix will be available?

**Comment by nicoddemus** [comment]
> Could you please let me know when the new release containing this fix will be available?

It will be available on the next release, you can subscribe for release notifications here on GitHub. :+1:

File: src/_pytest/tmpdir.py [hunk 15]
```diff
@@ -37,6 +35,73 @@
...
+        st = os.fstat(dir_fd)
+        if st.st_uid != user_id:
+            raise OSError(
+                f"The temporary directory {path} is not owned by the current user. "
+                "Fix this and try again."
+            )
+        if (st.st_mode & 0o077) != 0:
+            os.fchmod(dir_fd, st.st_mode & ~0o077)
+    finally:
+        os.close(dir_fd)
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: The docs describe the implementation, which is considered a bad practice.

Please move these remarks to comments in the body.


```suggestion
def _verify_ownership_and_tighten_permissions(path: Path, user_id: int) -> None:
    """Verify directory ownership and ensure private permissions (0o700),
    taking care to avoid TOCTOU races.

    Raises:
        OSError: If *path* cannot be safely opened (e.g. it is a symlink)
            or is not owned by *user_id*.
    """
    open_flags = os.O_RDONLY
    # Never follow symlinks (when these flags are available).
    for _flag in ("O_NOFOLLOW", "O_DIRECTORY"):
        open_flags |= getattr(os, _flag, 0)
    try:
        dir_fd = os.open(str(path), open_flags)
    except OSError as e:
        raise OSError(
            f"The temporary directory {path} could not be "
            "safely opened (it may be a symlink). "
            "Remove the symlink or directory and try again."
        ) from e
    try:
        # Use fd-based functions to eliminate TOCTOU races (CVE-2025-71176).
        st = os.fstat(dir_fd)
        if st.st_uid != user_id:
            raise OSError(
                f"The temporary directory {path} is not owned by the current user. "
                "Fix this and try again."
            )
        if (st.st_mode & 0o077) != 0:
            os.fchmod(dir_fd, st.st_mode & ~0o077)
    finally:
        os.close(dir_fd)
```

File: src/_pytest/tmpdir.py [hunk 16]
```diff
@@ -37,6 +35,73 @@
...
+    temproot: Path, prefix: str, keep: int, current: Path
+) -> None:
+    """Remove old randomly-named rootdirs, keeping the *keep* most recent.
+
+    *current* is excluded so the running session's rootdir is never removed.
+    Errors are silently ignored (other sessions may hold locks, etc.).
+
+    Uses ``os.scandir`` with ``follow_symlinks=False`` so that symlinks
+    planted under *temproot* are never followed during enumeration —
+    defense-in-depth against symlink attacks (CVE-2025-71176).
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: This is probably better as a comment in the code, as is describing an implementation detail.

File: src/_pytest/tmpdir.py [hunk 17]
```diff
@@ -155,40 +220,34 @@ def getbasetemp(self) -> Path:
...
             keep = self._retention_count
             if self._retention_policy == "none":
                 keep = 0
-            basetemp = make_numbered_dir_with_cleanup(
-                prefix="pytest-",
-                root=rootdir,
-                keep=keep,
-                lock_timeout=LOCK_TIMEOUT,
-                mode=0o700,
+            # keep-1 because the current session's rootdir (excluded from
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: 👍

File: src/_pytest/tmpdir.py (hunk 9 above)
Comment (nicoddemus) [reply src/_pytest/tmpdir.py]: Thanks, the extra directory seems to have been removed, but we still produce a symlink:

```python
def test(tmp_path): 
    print(tmp_path)
	

def test2(tmp_path): 
    print(tmp_path)
```

This produces:

```
ls /tmp/pytest-of-bruno-pw9tlny6
test0  test20  test2current  testcurrent
```

We will always have a single directory now, so the symlinks are no longer necessary.

File: changelog/14279.bugfix.rst [hunk 18]
```diff
@@ -0,0 +1,5 @@
+Temporary directories created by :fixture:`tmp_path` used to be created inside a ``pytest-of-{user}`` root directory within the system's temporary directory. Each session would create a new subdirectory with an incrementing counter -- ``pytest-0``, ``pytest-1``, etc. -- along with a ``pytest-current`` symlink pointing to the current one.
```
Comment (nicoddemus) [inline changelog/14279.bugfix.rst]: No need to duplicate the entry for the PR. This one can should be removed.

File: src/_pytest/pathlib.py [hunk 19]
```diff
@@ -159,9 +159,59 @@ def get_extended_length_path_str(path: str) -> str:
...
+        )
+
+
+def safe_rmtree(path: Path, *, ignore_errors: bool = False) -> None:
+    """Remove a directory tree with protection against symlink attacks.
+
+    Verifies that ``shutil.rmtree.avoids_symlink_attacks`` is True (the
+    platform provides a symlink-attack-resistant implementation) before
+    proceeding.  On platforms without this guarantee an explicit symlink
+    check is performed and a warning is emitted.
```
Comment (nicoddemus) [inline src/_pytest/pathlib.py]: Describing the implementation, better as  comment in the body.

File: src/_pytest/pathlib.py [hunk 20]
```diff
@@ -159,10 +159,53 @@ def get_extended_length_path_str(path: str) -> str:
...
+        raise OSError(
+            f"Refusing to recursively remove {path}: "
+            "path is a symlink, not a real directory."
+        )
+    # When `avoids_symlink_attacks` is True the platform's rmtree uses fd-based
+    # operations that are inherently resistant to symlink races — we only need
+    # to verify *path* itself is not a symlink (done above).
+    # When it is False a TOCTOU window remains for contents *inside* the tree,
+    # so we emit a one-time warning.
+    if not shutil.rmtree.avoids_symlink_attacks:
```
Comment (bluetech) [inline src/_pytest/pathlib.py:177]: I don't think that just because some poor soul happens to be using pytest on a system where `avoids_symlink_attacks` is false, we should warn them every time. There's nothing they can do about it after all.

File: src/_pytest/pathlib.py [hunk 21]
```diff
@@ -159,10 +159,53 @@ def get_extended_length_path_str(path: str) -> str:
     return long_path_prefix + path
 
 
+def _check_symlink_attack_safety(path: Path, *, stacklevel: int = 3) -> None:
+    """Guard against symlink attacks before recursive directory removal.
+
+    Raises ``OSError`` if *path* is a symlink.
+    """
+    if path.is_symlink():
```
Comment (bluetech) [inline src/_pytest/pathlib.py:167]: For me `os.shutil` already refuses a symlink argument. Is this check needed on our side then?

File: src/_pytest/tmpdir.py [hunk 22]
```diff
@@ -155,41 +233,37 @@ def getbasetemp(self) -> Path:
...
-            # we are the owners, to prevent any funny business (on unix, where
-            # temproot is usually shared).
-            # Also, to keep things private, fixup any world-readable temp
-            # rootdir's permissions. Historically 0o755 was used, so we can't
-            # just error out on this, at least for a while.
+                # getuser() likely returned illegal characters for the
+                # platform, fall back to a safe prefix.
+                rootdir_prefix = "pytest-of-unknown-"
+                rootdir = Path(tempfile.mkdtemp(prefix=rootdir_prefix, dir=temproot))
+            # Defense-in-depth: verify ownership and tighten permissions
```
Comment (bluetech) [inline src/_pytest/tmpdir.py:259]: Seems excessive to me. `mkdtemp` documentation already says "Creates a temporary directory in the most secure manner possible", why do we need to double check it?

File: src/_pytest/tmpdir.py [hunk 23]
```diff
@@ -155,41 +233,37 @@ def getbasetemp(self) -> Path:
...
             keep = self._retention_count
             if self._retention_policy == "none":
                 keep = 0
-            basetemp = make_numbered_dir_with_cleanup(
-                prefix="pytest-",
-                root=rootdir,
-                keep=keep,
-                lock_timeout=LOCK_TIMEOUT,
-                mode=0o700,
+            _cleanup_old_rootdirs(
```
Comment (bluetech) [inline src/_pytest/tmpdir.py:274]: The previous code comment said "use a sub-directory in the temproot to speed-up make_numbered_dir() call", and that's relevant here -- now this call will do a full scandir of `/tmp` on every cleanup, and the problem is that `/tmp` can accumulate a bunch of garbage via various programs, and seems like the previous code tried to avoid scanning it by using a sub-directory.

Comment (nicoddemus) [reply src/_pytest/tmpdir.py:274]: Good catch. 👍 

This PR now creates this structure:

```
/tmp/
  pytest-of-user-kjnasjnd/
  pytest-of-user-deadbeef/
  pytest-of-user-basu1234/
  pytest-of-user-olsdanj1/
```

If instead it created this:

```
/tmp/
  pytest-of-user/
    kjnasjnd/
    deadbeef/
    basu1234/
    olsdanj1/
```

Would that still work to prevent the security concerns this PR addresses?

I'm guessing the fact that an attack can create files in `/tmp/pytest-of-user` would still be a problem?

File: src/_pytest/pathlib.py (hunk 20 above)
Comment (orlitzky) [reply src/_pytest/pathlib.py:177]: I think all recent (>= POSIX 2008) platforms should be safe, so maybe it is a better idea to bail out here rather than run an unsafe recursive delete on the user's behalf? (In other words, let them do it, so pytest can't be blamed.) With the understanding that this will rarely happen on real systems.

**Comment by bluetech** [comment]
Hi, thanks for the PR. In the end we want with a simpler fix that preserves current behavior. If there's a need we can go with a more complex fix, but hopefully we can keep it simple.
